### ISSUE TITLE
Regression with TypeVar being used on Callable on 1.1.309

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#5128]
```python
from typing import Callable, TypeVar

_T = TypeVar("_T")

foo: Callable[[_T], _T]
```

Running pyright on that example will produce:

```
/tmp/test.py
  /tmp/test.py:5:16 - error: Type variable "_T" has no meaning in this context (reportGeneralTypeIssues)
  /tmp/test.py:5:21 - error: Type variable "_T" has no meaning in this context (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
A TypeVar must be bound to a function (in a `def` statement), a class (in a `class` statement) or a type alias. In your example, it's not bound to any of the above, so `_T` has no meaning in this case. Previously, pyright was missing a check here, so it failed to report an error in this condition. I don't consider this a regression. It's now working as it should.

In your code, is `foo` a local variable or an instance variable? If it's an instance variable, does the containing class define `_T` as a class-level type parameter?

The `Callable` special form doesn't provide a way to bind a type variable to it. You could use a callback protocol if that's your intent.

```python
from typing import Protocol, TypeVar

_T = TypeVar("_T")

class Identity(Protocol):
    def __call__(self, __val: _T) -> _T:
        ...

foo: Identity
```

**Comment by bellini666** [comment]
Fair enough! But when bounding it to both a function or a lambda it gives the same error:

```python
foo: Callable[[_T], _T] = lambda x: x
```
or
```python
def some_non_typed_func(x):
    return x

foo: Callable[[_T], _T] = some_non_typed_func
```

**Comment by erictraut** [comment]
The problem is the use of `_T` in the type annotation `Callable[[_T], _T]`. The value you assign to variable `foo` is irrelevant. The error will still be reported if you eliminate the assignment entirely as in your original code sample.

I think you might misunderstand what I mean by "binding". I'm not talking here about an assignment. I'm talking about associating a TypeVar with a scope (a function, class or TypeAlias). Refer to PEP 484 for details about TypeVar scopes.

**Comment by bellini666** [comment]
> I think you might misunderstand what I mean by "binding". I'm not talking here about an assignment. I'm talking about associating a TypeVar with a scope (a function, class or TypeAlias). Refer to PEP 484 for details about TypeVar scopes.

Oh yeah, I totally misunderstood it.

But ok, np then! Thought it was a regression because I had one code doing a cast in the return value of an overload function. I had to add that cast for older pyright versions or else it would not like it (even though it was the implementation and not the overloads, and it was marked as returning `Any`), and now that same code produced the error in this issue. But removing the cast did solve the issue, so probably what changed here was actually a fix to that issue I was having before
